### PR TITLE
Pass SID into store_info

### PR DESCRIPTION
**Pull request by arcusfelis**
This PR addresses MIM-1875 mongoose_session:merge_info merges data from different sessions

Proposed changes include:
* Pass SID into store_info
* Tweaking tests
* Remove merge_info function

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/172841) / small_tests / 0fe7d9ae5cb087ca3862603194229d541a842adb
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4007/172841/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4007/172841/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-04-10_20.50.42/lib.mongooseim.logs/run.2023-04-10_20.50.42/suite.log.html)

File: test/ejabberd_sm_SUITE.erl [hunk 1]
```diff
@@ -531,63 +581,6 @@ is_redis_running() ->
             false
     end.
 
-try_to_reproduce_race_condition(Config) ->
```
Comment (arcusfelis) [inline test/ejabberd_sm_SUITE.erl:534]: delete_session could only be deleted by process from SID.
same with set_session. So, the testcase is not needed.

Though there is a case when a session is removed during netsplit detection, but it is a different testcase.

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/172979) / small_tests / b968a3caeb2b434a1d098a13717271625c4aabba
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4007/172979/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4007/172979/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-04-12_17.32.17/lib.mongooseim.logs/run.2023-04-12_17.32.18/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/173367) / small_tests / 36b23f940ff3c27bac2cd76c20a561eea753f268
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4007/173367/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4007/173367/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-04-17_19.10.38/lib.mongooseim.logs/run.2023-04-17_19.10.39/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/173394) / small_tests / da7223118df03e0e34b2c2e18409a8afc37b19b5
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4007/173394/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4007/173394/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-04-17_20.59.14/lib.mongooseim.logs/run.2023-04-17_20.59.14/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/173621) / small_tests / 8e0c27611face59151b2d46e3bec4abe25511ddc
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4007/173621/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4007/173621/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-04-19_18.10.04/lib.mongooseim.logs/run.2023-04-19_18.10.04/suite.log.html)

File: src/stream_management/mod_stream_management.erl [hunk 2]
```diff
@@ -284,6 +284,7 @@ foreign_event(Acc, #{c2s_data := StateData,
                      event_type := {call, From},
                      event_tag := ?MODULE,
                      event_content := {resume, H}}, _Extra) ->
+    %% XXX Do we deregister from SM correctly?
```
Comment (chrzaszcz) [inline src/stream_management/mod_stream_management.erl:287]: This comment looks like a debug leftover, especially the "XXX". Please rewrite or remove.

File: test/ejabberd_sm_SUITE.erl [hunk 3]
```diff
@@ -44,19 +44,19 @@ tests() ->
      session_is_updated_when_created_twice,
      delete_session,
      clean_up,
-     too_much_sessions,
+     too_many_sessions,
```
Comment (chrzaszcz) [inline test/ejabberd_sm_SUITE.erl:47]: :+1:

File: test/ejabberd_sm_SUITE.erl [hunk 4]
```diff
@@ -181,6 +185,27 @@ session_info_is_updated_if_keys_match(C) ->
     [#session{sid = Sid, info = #{key1 := val2}}]
      = ?B(C):get_sessions(U,S).
 
+%% Same resource but different sids
+session_info_is_updated_properly_if_session_conflicts(C) ->
+    %% We use 2 seeds here
```
Comment (chrzaszcz) [inline test/ejabberd_sm_SUITE.erl:190]: I think it's "SIDs", not "seeds".

File: test/ejabberd_sm_SUITE.erl [hunk 5]
```diff
@@ -375,15 +395,34 @@ unload_meck() ->
...
+    do_meck_c2s(MeckC2s).
+
+do_meck_c2s(false) ->
+    ok;
+do_meck_c2s(true) ->
+    %% Very simple mock, not even reproducing async behaviour
+    %% It is for a limited use only, for more complex tests use a real c2s process (i.e. probably big tests)
+    meck:new(mongoose_c2s, [passthrough]),
+    meck:expect(mongoose_c2s, async_with_state, fun(Pid, Fun, Args) -> apply(Fun, [{ministate, Pid}|Args]), ok end),
+    GF = fun({ministate, Pid}) -> case ets:lookup(test_c2s_info, Pid) of [] -> #{}; [{Pid, Info}] -> Info end end,
```
Comment (chrzaszcz) [inline test/ejabberd_sm_SUITE.erl:419]: Please don't inline nested expressions like this. A named helper function might be better here.

File: test/ejabberd_sm_SUITE.erl [hunk 6]
```diff
@@ -181,6 +185,27 @@ session_info_is_updated_if_keys_match(C) ->
...
+    %% Sid2 > Sid in this case, because SIDs have a timestamp in them
+    %% We cannot test store_info for Sid2 though, because it has a different pid
+    Sid2 = make_sid_with_unique_pid(),
+
+    %% Two sessions for the same USR are registered after that:
+    given_session_opened(Sid, USR, 1, [{key1, val1}, {key2, a}]),
+    given_session_opened(Sid2, USR, 1, [{key1, val2}, {key3, b}]),
+
+    %% Overwritten without merging
+    %% Because we don't call open_session twice for the same Sid, so no need to merge info
```
Comment (chrzaszcz) [inline test/ejabberd_sm_SUITE.erl:201]: This comment is hard to understand, because we do call it for `Sid` twice: in lines 197 and 202.

**Review by chrzaszcz — COMMENTED** [review]
Looks good in general, I added some minor comments.

File: src/jingle_sip/jingle_sip_callbacks.erl [hunk 7]
```diff
@@ -91,6 +94,7 @@ translate_and_deliver_invite(Req, FromJID, FromBinary, ToJID, ToBinary) ->
     {ok, ReqID} = nksip_request:get_handle(Req),
 
     {CodecMap, SDP} = nksip_sdp_util:extract_codec_map(Body),
+    assert_sdp_record(Body, SDP),
```
[the remark this answers is not in the document]
Comment (chrzaszcz) [reply src/jingle_sip/jingle_sip_callbacks.erl:97]: It's fine, thanks for the explanation.

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/173710) / small_tests / 31a811636234a62540c5e6a8596c2e4383840a31
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4007/173710/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4007/173710/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-04-21_18.44.57/lib.mongooseim.logs/run.2023-04-21_18.44.58/suite.log.html)

**Pull request merged**